notebooks/writing_a_trace_transform_cpu_offloading.ipynb

```diff
@@ -0,0 +1,873 @@
+{
```

Comment:
<pre class="ql-syntax" spellcheck="false">Line #11.            # Smaller configuration
</pre><p>Question: To run Llama-3-8B w/o this tweak, how much GPU memory would we need?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/Lightning-AI/lightning-thunder/pull/966/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='22'/>

Reply:
With `thunder.examine.memory_caculation.get_alloc_memory` the estimated peak memory usage with forward is 3312.49 GB (note that this is with batchsize 8). Even with batch size of 1, the estimate is 442.23 GB.

For current configuration the estimate is 50.85 GB while actual is ~47GB.

```python
import torch
import thunder
from litgpt import Config, GPT
from functools import partial
from torch.testing import make_tensor

import thunder.examine
import thunder.examine.memory_caculation

N_LAYER = 7
BLOCK_SIZE = 1024

def get_model_and_args(batchdims=8):
    with torch.device("meta"):
        cfg: Config = Config.from_name("Llama-3-8B")
        # Smaller configuration
        # cfg.n_layer = N_LAYER
        # cfg.block_size = BLOCK_SIZE

        model = GPT(cfg)
        make = partial(make_tensor, low=0, high=255, device='meta', dtype=torch.int64, requires_grad=False)
        shape = (batchdims,) + (cfg.block_size,)

        x = make(shape)
        args, kwargs = (x,), {}

    return model, args, kwargs, cfg

model, args, kwargs, cfg = get_model_and_args()

jmodel = thunder.jit(model)

jmodel(*args, **kwargs)

fwd_trace = thunder.last_traces(jmodel)[-1]
memory_usage = thunder.examine.memory_caculation.get_alloc_memory(fwd_trace)
print(memory_usage[0] / 1e9, "GB")

```